File: ui/settings_text_area.go
package ui

import (
	"ily.dev/domi"
	"ily.dev/domi/html"

	"ily.dev/act3/ui/stimulus"
)

// SettingsTextArea renders an inline-editing textarea showing value.
// When the user commits an edit (blur after a change), commit is
// called with the new value and the resulting message is delivered.
// Escape reverts the textarea to value.
func SettingsTextArea[Msg any](value string, commit func(value string) Msg, attrs ...domi.Attr) domi.Node {
	return html.Div(
		Class("u-settings-text-area"),
		stimulus.Controller("settings-text-area"),
		group(attrs...),
	)(
		html.Textarea(
			Class("u-settings-text-area-input"),
			onChangeValue(commit),
			stimulus.Target("settings-text-area", "input"),
			stimulus.Action("keydown->settings-text-area#keydown"),
		)(domi.Text(value)),
	)
}

File: ui/settings-text-area.js
import { Controller } from "../web/stimulus.js";

// Presentation-only companion to the settings textarea. Saving rides
// the textarea's change event through domi; this controller only
// handles Escape, reverting to the last server-rendered value (the
// textarea's default value).
export default class extends Controller {
	static targets = ["input"];

	keydown(ev) {
		if (ev.key === "Escape") {
			this.inputTarget.value = this.inputTarget.defaultValue;
			this.inputTarget.blur();
		}
	}
}
